lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1313,6 +1313,57 @@ bool WitnessChecker::checkWitnessAccess(ValueDecl *requirement,
   return false;
 }
 
+/// If the getter or setter in a requirement is not marked as throws, but the
+/// witness specifies it as throws, then return true so it can be diagnosed.
+bool WitnessChecker::checkWitnessThrowingAccessor(ValueDecl *requirement,
+                                                  ValueDecl *witness) {
+  if (!isa<AbstractStorageDecl>(requirement) &&
+      !isa<AbstractStorageDecl>(witness)) {
+    return false;
+  }
+
+  auto shouldDiagnose = false;
+
+  auto reqStorageDecl = cast<AbstractStorageDecl>(requirement);
+  auto witnessStorageDecl = cast<AbstractStorageDecl>(witness);
+
+  auto requirementGetter = reqStorageDecl->getGetter();
+  auto witnessGetter = witnessStorageDecl->getGetter();
+
+  if (!requirementGetter->hasThrows() && witnessGetter->hasThrows()) {
+    shouldDiagnose = true;
+  }
+
+  if (requirement->isSettable(DC)) {
+    auto requirementSetter = reqStorageDecl->getSetter();
+    auto witnessSetter = witnessStorageDecl->getSetter();
+
+    if (!requirementSetter->hasThrows() && witnessSetter->hasThrows()) {
+      shouldDiagnose = true;
+    }
+  }
+
+  if (auto reqReadCoroutine = reqStorageDecl->getReadCoroutine()) {
+    if (auto witnessReadCoroutine = witnessStorageDecl->getReadCoroutine()) {
+      if (!reqReadCoroutine->hasThrows() && witnessReadCoroutine->hasThrows()) {
+        shouldDiagnose = true;
+      }
+    }
+  }
+
+  if (auto reqModifyCoroutine = reqStorageDecl->getModifyCoroutine()) {
+    if (auto witnessModifyCoroutine =
+            witnessStorageDecl->getModifyCoroutine()) {
+      if (!reqModifyCoroutine->hasThrows() &&
+          witnessModifyCoroutine->hasThrows()) {
+        shouldDiagnose = true;
+      }
+    }
+  }
+
+  return shouldDiagnose;
+}
```

Comment:
This should all be based on the `doesReadThrow()` etc. bits on the storage declaration.

Reply:
@rjmccall Done!